<?xml version="1.0"?>
<robot name="sensor_asus_xtion_pro" xmlns:xacro="http://ros.org/wiki/xacro">

     <xacro:include filename="$(find gazebo_sensor_collection)/urdf/includes/materials.urdf.xacro"/>

     <xacro:property name="asus_xtion_scale" value="1.0"/>
     <xacro:property name="asus_xtion_pro_depth_rel_rgb_py" value="0.0270" />
     <xacro:property name="asus_xtion_pro_rgb_rel_rgb_py"   value="-0.0220" />
     <xacro:property name="M_PI" value="3.1415926535897931" />

     <xacro:macro name="sensor_asus_xtion_pro" params="name parent *origin color:='white'">

          <joint name="${name}_rgb_joint" type="fixed">
               <xacro:insert_block name="origin" />
               <parent link="${parent}"/>
               <child link="${name}_rgb_frame" />
          </joint>

          <link name="${name}_rgb_frame">
               <inertial>
                    <mass value="0.001" />
                    <origin xyz="0 0 0" />
                    <inertia ixx="0.0001" ixy="0.0" ixz="0.0" iyy="0.0001" iyz="0.0" izz="0.0001" />
               </inertial>
          </link>

          <joint name="${name}_rgb_optical_joint" type="fixed">
               <origin xyz="0 0 0" rpy="${-M_PI/2} 0 ${-M_PI/2}" />
               <parent link="${name}_rgb_frame" />
               <child link="${name}_rgb_optical_frame" />
          </joint>

          <link name="${name}_rgb_optical_frame">
               <inertial>
                    <mass value="0.001" />
                    <origin xyz="0 0 0" />
                    <inertia ixx="0.0001" ixy="0.0" ixz="0.0" iyy="0.0001" iyz="0.0" izz="0.0001" />
               </inertial>
          </link>

          <joint name="${name}_joint" type="fixed">
               <origin xyz="0 ${asus_xtion_pro_rgb_rel_rgb_py} 0" rpy="0 0 0"/>
               <parent link="${name}_rgb_frame"/>
               <child link="${name}_link"/>
          </joint>

          <link name="${name}_link">
               <visual>
                    <origin xyz="-0.01 0 0" rpy="${-M_PI/2} -${M_PI} ${-M_PI/2}"/>
                    <geometry>
                         <mesh filename="file://$(find gazebo_sensor_collection)/meshes/cameras/asus/asus_xtion_pro_live.dae" scale="${asus_xtion_scale} ${asus_xtion_scale} ${asus_xtion_scale}"/>
                    </geometry>
                    <xacro:color_material color="${color}"/>
               </visual>
               <collision>
                    <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
                    <geometry>
                         <box size="0.0730 .2760 0.0720"/>
                    </geometry>
               </collision>
               <inertial>
                    <mass value="0.01" />
                    <origin xyz="0 0 0" />
                    <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001" />
               </inertial>
          </link>

          <joint name="${name}_depth_joint" type="fixed">
               <origin xyz="0 ${asus_xtion_pro_depth_rel_rgb_py} 0" rpy="0 0 0" />
               <parent link="${name}_rgb_frame" />
               <child link="${name}_depth_frame" />
          </joint>

          <link name="${name}_depth_frame">
               <inertial>
                    <mass value="0.01" />
                    <origin xyz="0 0 0" />
                    <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001" />
               </inertial>
          </link>

          <joint name="${name}_depth_optical_joint" type="fixed">
               <origin xyz="0 0 0" rpy="${-M_PI/2} 0 ${-M_PI/2}" />
               <parent link="${name}_depth_frame" />
               <child link="${name}_depth_optical_frame" />
          </joint>

          <link name="${name}_depth_optical_frame">
               <inertial>
                    <mass value="0.001" />
                    <origin xyz="0 0 0" />
                    <inertia ixx="0.0001" ixy="0.0" ixz="0.0" iyy="0.0001" iyz="0.0" izz="0.0001" />
               </inertial>
          </link>

          <!-- RGBD sensor for simulation, same as Kinect -->
          <xacro:sensor_xtion_pro_gazebo name="${name}"/>
          <xacro:color_gazebo parent="${name}_link" color="${color}"/>

     </xacro:macro>

     <!-- ====================================================
                           Gazebo
     ========================================================= -->

     <xacro:macro name="sensor_xtion_pro_gazebo" params="name">
          <gazebo reference="${name}_depth_frame">
               <sensor type="depth" name="${name}_sensor">
                    <always_on>true</always_on>
                    <update_rate>20.0</update_rate>
                    <camera>
                         <horizontal_fov>${60.0*M_PI/180.0}</horizontal_fov>
                         <image>
                              <format>R8G8B8</format>
                              <width>640</width>
                              <height>480</height>
                         </image>
                         <clip>
                              <near>0.05</near>
                              <far>3.5</far>
                         </clip>
                         <distortion>
                              <k1>0.0</k1>
                              <k2>0.0</k2>
                              <k3>0.0</k3>
                              <p1>0.0</p1>
                              <p2>0.0</p2>
                              <center>0.5 0.5</center>
                         </distortion>
                    </camera>

                    <plugin name="${name}_controller" filename="libgazebo_ros_camera.so">
                         <camera_name>${name}</camera_name>
                         <ros>
                              <!-- <namespace>custom_ns</namespace> -->
                              <remapping>${name}/image_raw:=${name}/rgb/image_raw</remapping>
                              <remapping>${name}/camera_info:=${name}/rgb/camera_info</remapping>
                              <remapping>${name}/image_depth:=${name}/depth/image_raw</remapping>
                              <remapping>${name}/camera_info_depth:=${name}/depth/camera_info</remapping>
                              <remapping>${name}/points:=${name}/depth/points</remapping>
                         </ros>

                         <frame_name>${name}_depth_optical_frame</frame_name>
                         <baseline>0.1</baseline>
                         <min_depth>0.4</min_depth>
                    </plugin>
               </sensor>
          </gazebo>
     </xacro:macro>
</robot>
